crates/criticalup-cli/src/commands/verify.rs

```diff
@@ -0,0 +1,98 @@
+// SPDX-FileCopyrightText: The Ferrocene Developers
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use std::{
+    env::current_dir,
+    os::unix::fs::MetadataExt,
+    path::{Path, PathBuf},
+};
+
+use criticaltrust::{integrity::IntegrityVerifier, signatures::Keychain};
+use criticalup_core::{
+    download_server_cache::DownloadServerCache,
+    download_server_client::DownloadServerClient,
+    project_manifest::{ProjectManifest, ProjectManifestProduct},
+    state::State,
+};
+use tracing::Span;
+use walkdir::WalkDir;
+
+use crate::{errors::Error, Context};
+
+#[tracing::instrument(level = "debug", skip_all, fields(project))]
+pub(crate) async fn run(
+    ctx: &Context,
+    manifest_path: Option<PathBuf>,
+    offline: bool,
+) -> Result<(), Error> {
+    let span = Span::current();
+    let manifest_path = if let Some(manifest_path) = manifest_path {
+        manifest_path
+    } else {
+        ProjectManifest::discover(&current_dir()?)?
+    };
+    span.record(
+        "manifest_path",
+        tracing::field::display(manifest_path.display()),
+    );
+
+    let state = State::load(&ctx.config).await?;
+    let maybe_client = if !offline {
+        Some(DownloadServerClient::new(&ctx.config, &state))
+    } else {
+        None
+    };
+    let cache = DownloadServerCache::new(&ctx.config.paths.cache_dir, &maybe_client).await?;
+    let keys = cache.keys().await?;
+
+    let project_manifest = ProjectManifest::load(&manifest_path)?;
+
+    let installation_dir = &ctx.config.paths.installation_dir;
+
+    for product in project_manifest.products() {
+        verify_product(&keys, installation_dir, product).await?;
+    }
+
+    Ok(())
+}
+
+#[tracing::instrument(level = "debug", skip_all, fields(product_path))]
+async fn verify_product(
+    keys: &Keychain,
+    installation_dir: &Path,
+    product: &ProjectManifestProduct,
+) -> Result<(), Error> {
+    let span = Span::current();
+
+    let mut integrity_verifier = IntegrityVerifier::new(keys);
+
+    let product_name = product.name();
+    let product_path = installation_dir.join(product.installation_id());
+    span.record(
+        "product_path",
+        tracing::field::display(product_path.display()),
+    );
+
+    tracing::info!("Verifying project `{product_name}`");
```

Comment:
```suggestion
    tracing::info!("Verifying product `{product_name}`");
```

We use `project` for user's project/criticalup.toml. I think you meant `product` here. Please correct me if I am mistaken.